src/http/server.ts

```diff
@@ -758,31 +760,246 @@ export class Server {
+    // `/users/:name?/:age?/:city?` should match  `/users`.
+    for (let i = 0; i < maxOptionalParams; i++) {
+      if (i === 0) { // We need to mark the start for the first optional param
+        // The below regex is very similar to `REGEX_URI_MATCHES` but this regex isn't
+        // global, and accounts for there being a required parameter before
+        tmpPath = tmpPath.replace(
+          /\/(:[^(/]+|{[^0-9][^}]*}\?)\/?/,
+          "/?([a-zA-Z0-9]+)?/?", // A `/` being optional, as well as the param being optional, and a ending `/` being optional
+        );
+      } else { // We can now create the replace regex for the rest taking into consideration the above replace regex
+        tmpPath = tmpPath.replace(
+          /\/?(:[^(/]+|{[^0-9][^}]*}\?)\/?/,
+          "([^/]+)?/?",
+        );
+      }
+    }
+
+    return {
+      og_path: path,
+      regex_path: `^${tmpPath}$`,
+      // Regex is same as other blocks, but we also strip the `?`.
+      params: (path.match(Server.REGEX_URI_MATCHES) || []).map(
+        (element: string) => {
+          return element.replace(/:|{|}|\?/g, "");
+        },
+      ),
+    };
+  }
+
+  /**
+   * Has the request URL path in question been handled previously? That is, a
+   * resource was found for it and an association was stored in the CACHED
+   * resource lookup Map.
+   *
+   * @param urlPath - The request URL path in question.
+   *
+   * @returns True if handled previously; false if not.
+   */
+  protected requestUrlWasHandledPreviously(urlPath: string): boolean {
+    return this.cached_resource_lookup_table.has(urlPath);
+  }
+
+  /**
+   * Get the lookup information on a resource given a request URL path and the
+   * resource index.
+   *
+   * @param urlPath - The request URL path to match to a resource.
+   * @param index - The index to search. This index contains associations
+   * between all regex paths that the server is handling and indices to
+   * resources in the resource lookup Map.
+   *
+   * @returns Lookup information on a resource or undefined if not found in the
+   * index that is used for the lookup information.
+   */
+  protected getResourceLookupInfo(
+    urlPath: string,
+  ): ISearchResult[] | undefined {
+    const url = urlPath.split("/");
+
+    if (url[url.length - 1] == "") {
+      url.pop();
+    }
+
+    if (url.length > 2) {
+      url.pop();
+    }
+
+    let urlWithoutParam = url.join("/");
+
+    if (urlWithoutParam.charAt(urlWithoutParam.length - 1) === "/") {
+      urlWithoutParam = urlWithoutParam.substring(
+        -1,
+        urlWithoutParam.length - 1,
+      );
+    }
+    const split = urlWithoutParam.split("/");
+    if (split.length > 1) {
+      if (urlWithoutParam[0] == "" || urlWithoutParam[0] == "/") {
+        urlWithoutParam = "/" + split[1];
+      } else {
+        urlWithoutParam = split[0];
       }
     }
+
+    try {
+      return this.resource_index_service!.getItem("\\^" + urlWithoutParam);
```

Comment:
this should return an object like the following

```
{
  id: 1,
  search_term: some regex path,
  search_input: some path,
  result: the value of the id in the lookup table
```

this prevents us from having to first find the index of an item and then taking that index and searching the lookup table. the index service should do all this for us. if we search for something, then it should return results from the lookup table and information about the search (the search input used and the id of the item in the lookup table, etc)

Reply:
5a694d2